source/common/common/callback_impl.h

```diff
@@ -81,5 +86,134 @@ template <typename... CallbackArgs> class CallbackManager {
+ *
+ * This is essentially a thread-safe version of the CallbackManager, but APIs to differ in some key
+ * regards. Specifically this class makes use of the Event::Dispatcher to coordinate events across
+ * threads and does involve some locking when it comes to updating the internally managed callback
+ * list.
+ *
+ * @note This is not safe to use for instances in which the lifetimes of the threads registering
+ * callbacks is less than the thread that owns the callback manager due to an assumption that
+ * dispatchers registered alongside callbacks remain valid, even if the callback expires.
+ *
+ * @see CallbackManager
+ */
+template <typename... CallbackArgs> class ThreadSafeCallbackManager {
+  struct CallbackHolder;
+  using CallbackListEntry = std::tuple<std::weak_ptr<CallbackHolder>, Event::Dispatcher&>;
+
+public:
+  using Callback = std::function<void(CallbackArgs...)>;
+
+  /**
+   * @param dispatcher Dispatcher relevant to the thread in which the callback manager is
+   * created/managed
+   */
+  explicit ThreadSafeCallbackManager(Event::Dispatcher& dispatcher) : dispatcher_(dispatcher) {}
+
+  /**
+   * @brief Add a callback.
+
+   * @param dispatcher Dispatcher from the same thread as the registered callback. This will be used
+   *                   to schedule the execution of the callback.
+   * @param callback callback to add
+   * @return ThreadSafeCallbackHandlePtr a handle that can be used to remove the callback.
+   */
+  ABSL_MUST_USE_RESULT ThreadSafeCallbackHandlePtr add(Event::Dispatcher& dispatcher,
+                                                       Callback callback) {
+    Thread::LockGuard lock(lock_);
+    auto new_callback = std::make_shared<CallbackHolder>(*this, callback);
+    callbacks_.push_back(CallbackListEntry(std::weak_ptr(new_callback), dispatcher));
+    // Get the list iterator of added callback handle, which will be used to remove itself from
+    // callbacks_ list.
+    new_callback->it_ = (--callbacks_.end());
+    return new_callback;
+  }
+
+  /**
+   * @brief Run all callbacks with a function that returns the input arguments
+   *
+   * NOTE: This code is currently safe if a callback deletes ITSELF from within a callback. It is
+   *       not safe if a callback deletes other callbacks.
+
+   * @param run_with function that is responsible for generating inputs to callbacks. This will be
+   * executed for each callback. Return values must be able to be safely copied across threads and
+   * asynchronously consumed.
+   */
+  void runCallbacksWith(std::function<std::tuple<CallbackArgs...>(void)> run_with) {
+    Thread::LockGuard lock(lock_);
+    for (auto it = callbacks_.cbegin(); it != callbacks_.cend();) {
+      auto& [cb, cb_dispatcher] = *(it++);
+
+      // sanity check cb is valid before attempting to schedule a dispatch
+      if (cb.expired()) {
+        continue;
+      }
+
+      auto args = run_with();
+      cb_dispatcher.post([cb = cb, args = args] {
```

Comment:
It's possible that I've missed something in my mental model, so I'll try to explain my mental model below because I don't believe what you're explaining is accurate, but please do correct me where I'm wrong.

---

The on-drain callback is registered per `Http::ConnectionManagerImpl` (HCM) instance. There is only one HCM instance per filter-chain, per-thread. In my current experience operating our mesh, the biggest I've seen is a 32 thread Envoy running 2 listeners each with 1 HTTP connection-manager defined in a single filter-chain. These machines see 10's of thousands of concurrent connections, but given this change I would only expect to see 64 registered drain callbacks (32 threads * 2 filter-chains * 1 HCM). 

Of course I can imagine running with a larger concurrency and configuring many more HTTP connection managers (and filter chains), but I am not connecting what sounds like a concern in the number of registered (and executed) callbacks and the number of open connections.

Reply:
My understanding is that there is one FilterManagerImpl per Network::ConnectionImpl.  ConnectionManagerImpl per FilterManagerImpl also implies that there's one per downstream connection.  I do find this somewhat confusing and counter intuitive.

I think that code like this https://github.com/envoyproxy/envoy/blob/4d2e0187d2dd4136c57c0f47cf5a209bcc13bda4/source/common/http/conn_manager_impl.cc#L812 which goes from ConnectionManagerImpl instance to Network::Connection instance supports my impression about the relationships between these objects and number of HCMs per Envoy process.

Reply:
Ahhh okay. This makes a lot more sense with the initial concerns then. Thank you for helping me form the right mental model! The comments on the original issue regarding batching also make more sense now. I'll spend some time figuring out how to batch these and/or avoid so many cross-thread calls.

Reply:
This has been addressed in the current iteration, making use of TLS to reduce the cross-thread interaction and I have also made callback registration a _non-thread-safe_ operation to simplify the implementation.